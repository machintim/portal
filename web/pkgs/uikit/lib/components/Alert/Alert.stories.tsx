import type { Meta, StoryObj } from '@storybook/react';
import Alert from './Alert';
import Link from '../Link';

const meta: Meta<typeof Alert> = {
  title: 'Components/Ready to use/Alert',
  component: Alert,
  argTypes: {
    type: {
      control: { type: 'select' },
      options: ['info', 'success', 'warning', 'error'],
    },
    testId: {
      control: 'text',
    },
  },

  decorators: [
    (Story) => (
      <div style={{ width: '350px' }}>
        <Story />
      </div>
    ),
  ],
};

export default meta;

type Story = StoryObj<typeof Alert>;

export const InfoAlert: Story = {
  render: (args) => (
    <div style={{ width: '300px' }}>
      <Alert {...args}>
        <Alert.Message>This is the information alert</Alert.Message>
      </Alert>
    </div>
  ),
  args: {
    type: 'info',
  },
};

export const InfoWithMessageAndTitleAlert: Story = {
  render: (args) => (
    <Alert {...args}>
      <Alert.Title>Info Alert</Alert.Title>
      <Alert.Message>This is the information alert</Alert.Message>
    </Alert>
  ),
};

export const InfoWithMessageTitleAndLinkAlert: Story = {
  render: (args) => (
    <Alert {...args}>
      <Alert.Title>Info Alert</Alert.Title>
      <div style={{ display: 'flex', alignItems: 'center' }}>
        <Alert.Message>
          This is the information alert{' '}
          <Link href='/troubleshoot'>link</Link>{' '}
        </Alert.Message>
      </div>
    </Alert>
  ),
};

export const ErrorAlert: Story = {
  render: (args) => (
    <Alert {...args}>
      <Alert.Title>Error Alert</Alert.Title>
      <Alert.Message>
        The operation was an error.{' '}
        <Link href='/troubleshoot'>Read more</Link>{' '}
      </Alert.Message>
    </Alert>
  ),
  args: {
    type: 'error',
  },
};

export const SuccessAlert: Story = {
  render: (args) => (
    <Alert {...args}>
      <Alert.Title>Success Alert</Alert.Title>
      <Alert.Message>
        The operation was successful.
        <Link href='/details'>View details</Link>
      </Alert.Message>
    </Alert>
  ),
  args: {
    type: 'success',
  },
};

export const WarningAlert: Story = {
  render: (args) => (
    <Alert {...args}>
      <Alert.Title>Warning Alert</Alert.Title>
      <Alert.Message>Proceed with caution!</Alert.Message>
    </Alert>
  ),
  args: {
    type: 'warning',
  },
};

export const CustomWidthAlert: Story = {
  render: (args) => (
    <div style={{ width: '500px' }}>
      <Alert {...args}>
        <Alert.Title>Custom Width Alert</Alert.Title>
        <Alert.Message>This alert has a custom width of 500px</Alert.Message>
      </Alert>
    </div>
  ),
};
