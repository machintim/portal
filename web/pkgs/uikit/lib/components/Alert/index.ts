export { default } from './Alert';
export { default as AlertTitle } from './AlertTitle';
export { default as AlertMessage } from './AlertMessage';
export type { AlertProps } from './Alert';
export type { AlertTitleProps } from './AlertTitle';
export type { AlertMessageProps } from './AlertMessage';
