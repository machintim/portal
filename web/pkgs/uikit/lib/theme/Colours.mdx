import { Meta, Title, Subtitle, ColorPalette, ColorItem } from '@storybook/addon-docs/blocks';
import theme from './defaultTheme';

<Meta title="theme/Colours" />
<Title>Colours explorer</Title>
<Subtitle>Colours approved for use in the UCL Design System</Subtitle> 

More information on the use of colour in the UCL Design System <a href="https://www.figma.com/design/8Sm5PxWOWJYpYXAzhRzUzt/UCL-Design-System-UI-Kit?node-id=1-287" target="_blank">can be found on Figma</a>. 

The colours populating this page are imported from the <a href="https://github.com/ucl-isd/uikit-react/blob/main/lib/theme/defaultTheme.ts#L5" target="_blank">theme file</a>. 

Below are all colours per category, one category per line: 

<ColorPalette>
{
  Object.entries(theme.color).map(([key, value]) => {
    return (
      <ColorItem key={key} title={key.charAt(0).toUpperCase() + key.slice(1)} colors={value} />
    );
  })
}
</ColorPalette>

Below is a breakdown of each colour category. These are the same colours as above, but grouped by category, one colour per line: 

{Object.entries(theme.color).map(([key, value]) => {
  return (
    <div key={key}>
      <h2>{key.charAt(0).toUpperCase() + key.slice(1)}</h2>

      <ColorPalette>

      {Object.entries(value).map(([colorKey, colorValue]) => {
        return (
          <ColorItem key={colorKey} title={colorKey} colors={{ [colorKey]: colorValue }} />
        );
      })}

      {console.log(key, value)}

      </ColorPalette>
    </div>
  );
})}
