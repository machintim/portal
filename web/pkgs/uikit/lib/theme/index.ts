export { default as theme } from './defaultTheme';
export type {
  ThemeType,
  DisplayColoursType,
  BlackAndWhiteColoursType,
} from './defaultTheme';
export {
  default as useTheme,
  ThemeContextProvider,
} from './useTheme';
