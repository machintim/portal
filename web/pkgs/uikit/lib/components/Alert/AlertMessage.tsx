import { css, cx } from '@emotion/css';
import { theme } from '../../theme';
import { HTMLAttributes } from 'react';

export const NAME = 'ucl-uikit-alert__message';

export interface AlertMessageProps
  extends HTMLAttributes<HTMLParagraphElement> {
  testId?: string;
}

const AlertMessage = ({
  testId = NAME,
  className,
  children,
}: AlertMessageProps) => {
  const messageStyles = css`
    font-size: ${theme.font.size.f16};
    color: ${theme.color.text.primary};
    margin: 0;
    line-height: 1.5;
  `;
  const style = cx(messageStyles, NAME, className);

  return (
    <p
      className={style}
      data-testid={testId}
    >
      {children}
    </p>
  );
};

export default AlertMessage;
