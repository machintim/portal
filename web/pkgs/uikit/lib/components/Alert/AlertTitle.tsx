import { css } from '@emotion/css';
import { theme } from '../../theme';
import { HTMLAttributes } from 'react';

export const NAME = 'ucl-uikit-alert__title';

export interface AlertTitleProps
  extends HTMLAttributes<HTMLHeadingElement> {
  testId?: string;
}

const AlertTitle = ({
  children,
  testId = NAME,
}: AlertTitleProps) => {
  const titleStyles = css`
    margin: 0 0 ${theme.margin.m8} 0;
    font-family: ${theme.font.family.primary};
    font-size: ${theme.font.size.f16};
    font-weight: 700;
    color: ${theme.color.text.primary};
  `;

  return (
    <h3
      className={titleStyles}
      data-test-id={testId}
    >
      {children}
    </h3>
  );
};

export default AlertTitle;
