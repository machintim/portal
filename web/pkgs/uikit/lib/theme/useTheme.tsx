// We will probably move away from context/provider to store/hook for accessing theme.

import {
  createContext,
  useState,
  useContext,
  ReactNode,
} from 'react';

import theme, { ThemeType } from './defaultTheme';

const useThemeState = () => useState(theme as ThemeType);

const ThemeContext = createContext<ReturnType<
  typeof useThemeState
> | null>(null);

export const ThemeContextProvider: React.FC<{
  children: ReactNode;
}> = ({ children }) => (
  <ThemeContext.Provider value={useThemeState()}>
    {children}
  </ThemeContext.Provider>
);

const useTheme = () => {
  const value = useContext(ThemeContext);
  if (value === null) {
    throw new Error('Please add ThemeContextProvider');
  }
  return value;
};

export default useTheme;
