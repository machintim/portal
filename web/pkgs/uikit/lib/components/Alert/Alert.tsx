import { createContext, useContext, HTMLAttributes } from 'react';
import { css, cx } from '@emotion/css';
import { theme } from '../../theme';
import Icon from '../Icon';
import AlertMessage from './AlertMessage';
import AlertTitle from './AlertTitle';

export const NAME = 'ucl-uikit-alert';

export type AlertType = 'info' | 'success' | 'warning' | 'error';

export interface AlertContextProps {
  type: AlertType;
}

export const AlertContext = createContext<AlertContextProps | null>(null);

export function useAlertContext() {
  const context = useContext(AlertContext);
  if (!context) {
    throw new Error('useAlertContext must be used within an Alert');
  }
  return context;
}

export interface AlertProps extends HTMLAttributes<HTMLDivElement> {
  type: AlertType;
  testId?: string;
}

const Alert = ({
  type,
  children,
  testId = NAME,
  className,
  ...props
}: AlertProps) => {
  const baseStyles = css`
    display: flex;
    font-family: ${theme.font.family.primary};
    align-items: flex-start;
    gap: ${theme.margin.m8};
    padding: ${theme.padding.p16} ${theme.padding.p12} ${theme.padding.p16}
      ${theme.padding.p8};
    border-left-width: 4px;
    border-left-style: solid;
  `;

  const infoStyle = css`
    background-color: ${theme.color.interaction.blue10};
    border-color: ${theme.color.interaction.blue70};
  `;

  const successStyle = css`
    background-color: ${theme.color.system.green10};
    border-color: ${theme.color.system.green70};
  `;

  const warningStyle = css`
    background-color: ${theme.color.system.orange10};
    border-color: ${theme.color.system.orange70};
  `;

  const errorStyle = css`
    background-color: ${theme.color.system.red10};
    border-color: ${theme.color.system.red70};
  `;

  const style = cx(
    NAME,
    baseStyles,
    type === 'info' && infoStyle,
    type === 'success' && successStyle,
    type === 'warning' && warningStyle,
    type === 'error' && errorStyle,
    className
  );

  const iconStyle = cx(
    type === 'info' &&
      css`
        color: ${theme.color.interaction.blue70};
      `,
    type === 'success' &&
      css`
        color: ${theme.color.system.green70};
      `,
    type === 'warning' &&
      css`
        color: ${theme.color.system.orange70};
      `,
    type === 'error' &&
      css`
        color: ${theme.color.system.red70};
      `
  );

  const IconComp = {
    info: Icon.Info,
    success: Icon.CheckCircle,
    warning: Icon.AlertTriangle,
    error: Icon.XCircle,
  }[type];

  return (
    <AlertContext.Provider value={{ type }}>
      <div
        className={style}
        data-testid={testId}
        role='alert'
        {...props}
      >
        <IconComp className={iconStyle} />
        <div>{children}</div>
      </div>
    </AlertContext.Provider>
  );
};

export interface IAlertSubComponents {
  Title: typeof AlertTitle;
  Message: typeof AlertMessage;
}

const AlertWithSubComponents = Alert as typeof Alert & IAlertSubComponents;

Alert.Title = AlertTitle;
Alert.Message = AlertMessage;

export default AlertWithSubComponents;
