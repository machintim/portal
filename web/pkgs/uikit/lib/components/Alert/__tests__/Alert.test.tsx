import { describe, expect, test } from 'vitest';
import { render } from '@testing-library/react';
import { ThemeContextProvider } from '../../../theme/useTheme';
import Alert from '../Alert';

describe('Alert', () => {
  // Snapshot tests

  test('snapshot: type=info, no title', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <Alert type='info'>
          <Alert.Message>
            This is an informational alert.
          </Alert.Message>
        </Alert>
      </ThemeContextProvider>
    );
    expect(
      renderResult.container.firstChild
    ).toMatchSnapshot();
  });

  test('snapshot: type=success', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <Alert type='success'>
          <Alert.Title>Success</Alert.Title>
          <Alert.Message>
            Your operation completed successfully.
          </Alert.Message>
        </Alert>
      </ThemeContextProvider>
    );
    expect(
      renderResult.container.firstChild
    ).toMatchSnapshot();
  });

  test('snapshot: type=warning', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <Alert type='warning'>
          <Alert.Title>Warning</Alert.Title>
          <Alert.Message>
            This operation may have unintended consequences.
          </Alert.Message>
        </Alert>
      </ThemeContextProvider>
    );
    expect(
      renderResult.container.firstChild
    ).toMatchSnapshot();
  });

  test('snapshot: type=error', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <Alert type='error'>
          <Alert.Title>Error Alert</Alert.Title>
          <Alert.Message>
            The operation failed.
          </Alert.Message>
        </Alert>
      </ThemeContextProvider>
    );
    expect(
      renderResult.container.firstChild
    ).toMatchSnapshot();
  });
});
