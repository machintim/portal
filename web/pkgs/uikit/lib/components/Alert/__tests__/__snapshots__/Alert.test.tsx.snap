// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Alert > snapshot: type=error 1`] = `
<div
  class="ucl-uikit-alert css-1bgvdmk"
  data-testid="ucl-uikit-alert"
  role="alert"
>
  <svg
    class="ucl-icon css-nysdxe"
    data-testid="ucl-icon"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="12"
      cy="12"
      r="10"
    />
    <line
      x1="15"
      x2="9"
      y1="9"
      y2="15"
    />
    <line
      x1="9"
      x2="15"
      y1="9"
      y2="15"
    />
  </svg>
  <div>
    <h3
      class="css-1ybx1vr"
      data-test-id="ucl-uikit-alert__title"
    >
      Error Alert
    </h3>
    <p
      class="css-omeh0l ucl-uikit-alert__message"
      data-testid="ucl-uikit-alert__message"
    >
      The operation failed.
    </p>
  </div>
</div>
`;

exports[`Alert > snapshot: type=info, no title 1`] = `
<div
  class="ucl-uikit-alert css-1rd5b6h"
  data-testid="ucl-uikit-alert"
  role="alert"
>
  <svg
    class="ucl-icon css-1oh4ubr"
    data-testid="ucl-icon"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="12"
      cy="12"
      r="10"
    />
    <line
      x1="12"
      x2="12"
      y1="16"
      y2="12"
    />
    <line
      x1="12"
      x2="12.01"
      y1="8"
      y2="8"
    />
  </svg>
  <div>
    <p
      class="css-omeh0l ucl-uikit-alert__message"
      data-testid="ucl-uikit-alert__message"
    >
      This is an informational alert.
    </p>
  </div>
</div>
`;

exports[`Alert > snapshot: type=success 1`] = `
<div
  class="ucl-uikit-alert css-17ofdem"
  data-testid="ucl-uikit-alert"
  role="alert"
>
  <svg
    class="ucl-icon css-4xbsfp"
    data-testid="ucl-icon"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
    />
    <polyline
      points="22 4 12 14.01 9 11.01"
    />
  </svg>
  <div>
    <h3
      class="css-1ybx1vr"
      data-test-id="ucl-uikit-alert__title"
    >
      Success
    </h3>
    <p
      class="css-omeh0l ucl-uikit-alert__message"
      data-testid="ucl-uikit-alert__message"
    >
      Your operation completed successfully.
    </p>
  </div>
</div>
`;

exports[`Alert > snapshot: type=warning 1`] = `
<div
  class="ucl-uikit-alert css-1whwjpp"
  data-testid="ucl-uikit-alert"
  role="alert"
>
  <svg
    class="ucl-icon css-dwkbnw"
    data-testid="ucl-icon"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M10.29 3.86L1.82 18a2 2 0 0 0 1.71 3h16.94a2 2 0 0 0 1.71-3L13.71 3.86a2 2 0 0 0-3.42 0z"
    />
    <line
      x1="12"
      x2="12"
      y1="9"
      y2="13"
    />
    <line
      x1="12"
      x2="12.01"
      y1="17"
      y2="17"
    />
  </svg>
  <div>
    <h3
      class="css-1ybx1vr"
      data-test-id="ucl-uikit-alert__title"
    >
      Warning
    </h3>
    <p
      class="css-omeh0l ucl-uikit-alert__message"
      data-testid="ucl-uikit-alert__message"
    >
      This operation may have unintended consequences.
    </p>
  </div>
</div>
`;
